# lab 04 - Pandas 2
Name: Robb Alexander    
Class: CSCI349    
Semester: 2021SP    
Instructor: Brian King


In [1]:
# Setting things up
import sys
import numpy as np
import pandas as pd

In [2]:
days = ["Mon","Tue","Wed","Thurs","Fri","Student"]
scores = pd.DataFrame([pd.Series([9.5, 8.75, 8, 10, 7.75, "Bob"], index=days, name="week_1"), pd.Series([9, 8, 10, 8.75, 7.25, "Jane"], index=days, name="week_1"), pd.Series([8.5, 0, 9.75, 9, 6, "Bob"], index=days, name="week_2"), pd.Series([7, 8.25, 9.25, 0, 8, "Jane"], index=days, name="week_2"), pd.Series([7, 8.5, 9.25, 0, 0, "Bob"], index=days, name="week_3"), pd.Series([8.5, 8.25, 9, 8, 7.5, "Jane"], index=days, name="week_3"), pd.Series([9, 7, 8.5, 8, 8, "Bob"], index=days, name="week_4"), pd.Series([9.25, 9.25, 8.5, 7.75, 7.5, "Jane"], index=days,name="week_4")])

#### 1) Report the shape, the number of dimensions, the size of the data, and the data types of each column. Print each separately. Then create an additional code cell that shows the results of scores.info().

In [3]:
print("The shape of scores:", scores.shape)
print("The dimensions of scores:", scores.ndim)
print("The size of scores:", scores.size)
print("The data types of scores:", scores.dtypes)

The shape of scores: (8, 6)
The dimensions of scores: 2
The size of scores: 48
The data types of scores: Mon        float64
Tue        float64
Wed        float64
Thurs      float64
Fri        float64
Student     object
dtype: object


In [4]:
print(scores.info())

<class 'pandas.core.frame.DataFrame'>
Index: 8 entries, week_1 to week_4
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Mon      8 non-null      float64
 1   Tue      8 non-null      float64
 2   Wed      8 non-null      float64
 3   Thurs    8 non-null      float64
 4   Fri      8 non-null      float64
 5   Student  8 non-null      object 
dtypes: float64(5), object(1)
memory usage: 448.0+ bytes
None


#### 2) Show two different ways to report the number of observations in scores.

In [5]:
print(len(scores))
print(scores.shape[0])

8
8


#### 3) Show the variables (i.e. columns) in scores

In [6]:
print(list(scores.columns))

['Mon', 'Tue', 'Wed', 'Thurs', 'Fri', 'Student']


#### 4) There are repetitive names in the index. This happens quite a bit, depending on how the rows are to be named. For now. print the unique index names in scores

In [7]:
print(list(scores.index.unique()))

['week_1', 'week_2', 'week_3', 'week_4']


#### 5) Rename the 'Thurs' column header to be 'Thu' . (Hint: Changing scores.columns[3] directly will not work, not to mention it would be a poor approach to hard code a location this way! Magic numbers are never a good idea when referencing index or column locations. They may change!) Show the new scores data frame. This should be the same, except with Thurs changed to Thu.

In [8]:
scores = scores.rename(columns={"Thurs": "Thu"})

#### 6) Now, make a bigger change. Let's rename 'week_1' to be 'w1', 'week_2' to be 'w2', and so on. Also, suppose you decide to use only 2 letter abbreviations for the days. Rename the days in the column names to be 'Mo', 'Tu', 'We', 'Th', 'Fr'. Show the updated scores dataframe. It should appear as follows:

In [9]:
original = list(scores.columns)
new = [i[:2] for i in original if i != "Student"]
col_rename_map = {o:n for o, n in zip(original, new)}

original = list(scores.index)
new = ["w"+i.split("_")[-1] for i in original]
row_rename_map = {o:n for o, n in zip(original, new)}

scores = scores.rename(index=row_rename_map, columns=col_rename_map)

#### 7) Compare the type of the expression scores['Mo'] vs. scores[['Mo']]. What is the difference? (HINT: use the type() Python function on both.)

In [10]:
"""
The difference between the two is that with one 
 bracket is a Series,  while two is a DataFrame
"""
print(type(scores['Mo']).__name__)
print(type(scores[['Mo']]).__name__)

Series
DataFrame


#### 8) Demonstrate the describe() method on scores. What type does it return?

In [11]:
print(type(scores.describe()).__name__)
print(scores.describe())

DataFrame
             Mo      Tu         We         Th        Fr
count  8.000000  8.0000   8.000000   8.000000  8.000000
mean   8.468750  7.2500   9.031250   6.437500  6.500000
std    0.967669  3.0000   0.673974   4.037215  2.702512
min    7.000000  0.0000   8.000000   0.000000  0.000000
25%    8.125000  7.7500   8.500000   5.812500  6.937500
50%    8.750000  8.2500   9.125000   8.000000  7.500000
75%    9.062500  8.5625   9.375000   8.812500  7.812500
max    9.500000  9.2500  10.000000  10.000000  8.000000


#### 9) Store the output of describe() as a variable. Then, using this data frame, report the day that had the largest standard deviation in quiz scores and its value. Clean your temporary variables when complete (i.e. use del)

In [12]:
desc = scores.describe()
stds = desc.loc['std']
i = desc.loc['std'].argmax()

print("Day with largest standard deviation:", stds.index[i])
del desc

Day with largest standard deviation: Th


#### 10) Knowing that describe() is nothing more than another data frame is interesting. However, it's more important to write code to output the specific results you are looking for. Write the Python code to repeat the previous exercise without using describe().

In [13]:
functions = [(pd.DataFrame.count, None), (pd.DataFrame.mean, None), (pd.DataFrame.std, None), (pd.DataFrame.min, None), (pd.DataFrame.quantile, 0.25), (pd.DataFrame.quantile, 0.5), (pd.DataFrame.quantile, 0.75), (pd.DataFrame.max, None)]
data = []
for f, p in functions:
    if p:
        data.append(f(scores,q=p))
    else:
        data.append(f(scores))

desc = pd.DataFrame(data, index=[f[0].__name__ if f[0].__name__ != "quantile" else str(f[1] * 100)[:2] + "%"  for f in functions])
desc = desc.drop(['Student'], axis=1)
print(desc)

             Mo      Tu         We         Th        Fr
count  8.000000  8.0000   8.000000   8.000000  8.000000
mean   8.468750  7.2500   9.031250   6.437500  6.500000
std    0.967669  3.0000   0.673974   4.037215  2.702512
min    7.000000  0.0000   8.000000   0.000000  0.000000
25%    8.125000  7.7500   8.500000   5.812500  6.937500
50%    8.750000  8.2500   9.125000   8.000000  7.500000
75%    9.062500  8.5625   9.375000   8.812500  7.812500
max    9.500000  9.2500  10.000000  10.000000  8.000000


#### 11) Write the code that changes all 0.0 entries to np.nan. (NOTE: You should be able to do this with just one line of code using pandas selection techniques!) Then, show scores. All 0.0 entries should be replaced with NaN in the output.

In [14]:
scores = scores.replace(0, np.NaN)
print(scores)

      Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50   NaN   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25    NaN  8.00    Jane
w3  7.00  8.50   9.25    NaN   NaN     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane


#### 12) Show the output of describe() again. Yes, this is better! Note how much more informative this output is when properly dealing with missing data. Copy your code from above to output the day with the largest standard deviation in quiz scores. You should have a substantially different result than you did before!

In [15]:
new_desc = scores.describe()
print(new_desc)

stds = new_desc.loc['std']
i = new_desc.loc['std'].argmax()

print("Day with largest standard deviation:", stds.index[i])
del new_desc

             Mo        Tu         We         Th        Fr
count  8.000000  7.000000   8.000000   6.000000  7.000000
mean   8.468750  8.285714   9.031250   8.583333  7.428571
std    0.967669  0.698638   0.673974   0.846562  0.687906
min    7.000000  7.000000   8.000000   7.750000  6.000000
25%    8.125000  8.125000   8.500000   8.000000  7.375000
50%    8.750000  8.250000   9.125000   8.375000  7.500000
75%    9.062500  8.625000   9.375000   8.937500  7.875000
max    9.500000  9.250000  10.000000  10.000000  8.000000
Day with largest standard deviation: Mo


#### 13) Show the output of scores.values. What does the .values attribute do?

In [16]:
"""
values returns a ndarray 2D array which is the
 dataframe represented in a numpy form. It is
 recommended now to use DataFrame.to_numpy()
"""
print(scores.values)
print(type(scores.values).__name__)

[[9.5 8.75 8.0 10.0 7.75 'Bob']
 [9.0 8.0 10.0 8.75 7.25 'Jane']
 [8.5 nan 9.75 9.0 6.0 'Bob']
 [7.0 8.25 9.25 nan 8.0 'Jane']
 [7.0 8.5 9.25 nan nan 'Bob']
 [8.5 8.25 9.0 8.0 7.5 'Jane']
 [9.0 7.0 8.5 8.0 8.0 'Bob']
 [9.25 9.25 8.5 7.75 7.5 'Jane']]
ndarray


#### 14) What is the mean quiz score for each day? Do not use describe()

In [17]:
scores.mean()

Mo    8.468750
Tu    8.285714
We    9.031250
Th    8.583333
Fr    7.428571
dtype: float64

#### 15) What is the mean quiz score over the entire dataset, ignoring all missing values? (HINT: there are many ways to do this, and no solution should involve writing any loops! The only right answer is 8.375. If you aren't getting that, then you are not handling NaN values correctly. You only want the mean over all scores that are NOT NaN!)

In [18]:
count = scores.count().drop("Student").sum()
sums = scores.sum().drop("Student").sum()
print(sums / count)

8.375


#### 16) Show the mean for each week over all students. Again, you are not counting missing quizzes into the average. (Yes, technically you could write a loop to compute this exhaustively. However, you should rarely need to do that. Use pandas!)

In [19]:
count = scores.groupby(level=0).count().drop("Student", axis=1).sum(axis=1)
sums = scores.groupby(level=0).sum().sum(axis=1)
print(sums / count)

w1    8.70000
w2    8.21875
w3    8.25000
w4    8.27500
dtype: float64


#### 17) Show the mean score for each student. Again, ignore all missing values.

In [20]:
count = scores.groupby("Student").count().sum(axis=1)
sums = scores.groupby("Student").sum().sum(axis=1)

print(sums / count)

Student
Bob     8.382353
Jane    8.368421
dtype: float64


#### 18) Select the data frame for only Monday and Student variables using .loc, and again using .iloc

In [21]:
print(scores.loc[:,["Mo","Student"]])
print(scores.iloc[:,[0,-1]])

      Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane
      Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane


#### 19) Show a new DataFrame containing the week as an index (like the original), but contains only each student with the count of their quizzes and the mean of their scores. Your result should look as follows:

In [35]:
names = scores.loc[:,["Student"]]
count = scores.drop("Student", axis=1).count(axis=1)
mean = scores.drop("Student", axis=1).sum(axis=1) / count
names['Count'] = count
names['mean'] = mean

names

Student  Count    mean
w1     Bob      5  8.8000
w1    Jane      5  8.6000
w2     Bob      4  8.3125
w2    Jane      4  8.1250
w3     Bob      3  8.2500
w3    Jane      5  8.2500
w4     Bob      5  8.1000
w4    Jane      5  8.4500

#### 20) Show the number of times Friday's score was greater than 7.25 for each student. Start by selecting a subset of observations, then use value_counts() on the Student variable of the selected data. Your result should look as follows:

In [34]:
new_scores = scores.loc[:,["Fr","Student"]]
new_scores = new_scores[new_scores['Fr'] > 7.25]

new_scores["Student"].value_counts()

Jane    3
Bob     2
Name: Student, dtype: int64

#### 21) Repeat the previous exercise, but this time use the where() method on scores to select your data. You should have the same output.

In [33]:
new_scores = scores.loc[:,["Fr","Student"]]
new_scores = new_scores.where(new_scores['Fr'] > 7.25)
new_scores = new_scores.dropna()

new_scores["Student"].value_counts()

Jane    3
Bob     2
Name: Student, dtype: int64

#### 22) Select the scores that were greater than the mean score over the entire dataset (without the missing values). Again, no hard coding values here (i.e. do NOT hard code the mean, but recompute it. Then, report the number of scores for each day that exceeded that global mean. Your result should look like:

In [32]:
count = scores.count().drop("Student").sum()
sums = scores.sum().drop("Student").sum()
mean = sums / count

new_scores = scores.drop("Student", axis=1)
new_scores.where(new_scores > mean).count()

Mo    6
Tu    3
We    7
Th    3
Fr    0
dtype: int64

#### 23) When dealing with enormous datasets for modeling, we often drop all observations that have any missing data. Show a new data frame that has only complete observations.

In [30]:
scores.dropna()

Mo    Tu    We     Th    Fr Student
w1  9.50  8.75   8.0  10.00  7.75     Bob
w1  9.00  8.00  10.0   8.75  7.25    Jane
w3  8.50  8.25   9.0   8.00  7.50    Jane
w4  9.00  7.00   8.5   8.00  8.00     Bob
w4  9.25  9.25   8.5   7.75  7.50    Jane